lib/Basic/Assertions.cpp

```diff
@@ -0,0 +1,55 @@
+//===--- Assertions.cpp - Swift Version Number -------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines custom assertion support functions
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Basic/Assertions.h"
+#undef NDEBUG
+#include <cassert>
+#include <iostream>
+
+int CONDITIONAL_ASSERT_Global_enable_flag =
+#ifdef NDEBUG
+  0; // Default to `off` in release builds
+#else
+  0; // TODO: Default to `on` in debug builds
+#endif
+
+[[noreturn]] void ASSERT_failure(const char *expr, const char *file, int line, const char *func) {
+  // Only print the last component of `file`
+  const char *f = file;
+  for (const char *p = file; *p != '\0'; p++) {
+    if ((p[0] == '/' || p[0] == '\\')
+	&& p[1] != '/' && p[1] != '\\' && p[1] != '\0') {
+      f = p + 1;
+    }
+  }
+  // Format here matches that used by `assert` on macOS:
+  std::cerr
+    << "Assertion failed: "
+    << "(" << expr << "), "
+    << "function " << func << ", "
+    << "file " << f << ", "
+    << "line " << line << "."
+    << std::endl;
+  abort();
```

Comment:
It would be handy to have more options than (a) trap on first assertion failure (b) ignore all assertions, considering that discovering which assertions are incorrect will be a gradual process, and the first assertion that fails may not be related to the problem directly being solved.

For example:
- print assertions and backtraces that fail but keep running
- trap on the Nth assertion
- break on assertion failures but keep running (`LLVM_BUILTIN_DEBUGTRAP`)
- break on the Nth assertion failure

Reply:
All of that could be handled with future changes to this function, if we made the function declaration to not be `[[noreturn]]`.

Reply:
I added two very basic controls as an experiment:
```
-Xllvm -assert-help           Dumps a list of assertion control options the first time an assertion is hit
-Xllvm -assert-continue    Assertion failures are printed but don't abort()
```

If this general approach (using -Xllvm) works for folks, then this seems pretty easy to expand with whatever other controls we might like.   Another possibility would be to provide an interactive mode similar to what @al45tair did for backtracing:  An assertion would then print a prompt and wait for you to press a key.  You could choose to continue, drop into a debugger, or abort().  After a timeout, the default handling would proceed.

Reply:
Two sub-variants of the above could integrate better with complex builds: emit a warning on failure, and emit an error on failure (both continuing execution afterwards). But obviously does not have to be in the first version of this!

Reply:
Yes, I plan to keep the first version pretty basic:  There's a lot of potentially-disruptive work to merge this, and I want to be sure folks are comfortable with me roto-tilling all the assertions in the project.  Once that's done, there are many ways to improve how assertions actually get reported and processed at runtime.

Reply:
It might also be good to have a *selective* continue option, like:

```
-Xllvm -assert-work-around=TypeCheckAttr.cpp:181
```

This would provide a way to work around a specific assertion that's been investigated and found to be spurious for a particular project.